redsky/streams.py

```diff
@@ -0,0 +1,630 @@
+        self.buffer.append(x)
+        if len(self.buffer) == self.n:
+            return self.emit(tuple(self.buffer))
+        else:
+            return []
+
+
+class timed_window(Stream):
+    def __init__(self, interval, child, loop=None):
+        self.interval = interval
+        self.buffer = []
+        self.last = gen.moment
+
+        Stream.__init__(self, child, loop=loop)
+
+        self.loop.add_callback(self.cb)
+
+    def update(self, x, who=None):
+        self.buffer.append(x)
+        return self.last
+
+    @gen.coroutine
+    def cb(self):
+        while True:
+            L, self.buffer = self.buffer, []
+            self.last = self.emit(L)
+            yield self.last
+            yield gen.sleep(self.interval)
+
+
+class delay(Stream):
+    def __init__(self, interval, child, loop=None):
+        self.interval = interval
+        self.queue = Queue()
+
+        Stream.__init__(self, child, loop=loop)
+
+        self.loop.add_callback(self.cb)
+
+    @gen.coroutine
+    def cb(self):
+        while True:
+            last = time()
+            x = yield self.queue.get()
+            yield self.emit(x)
+            duration = self.interval - (time() - last)
+            if duration > 0:
+                yield gen.sleep(duration)
+
+    def update(self, x, who=None):
+        return self.queue.put(x)
+
+
+class rate_limit(Stream):
+    def __init__(self, interval, child):
+        self.interval = interval
+        self.next = 0
+
+        Stream.__init__(self, child)
+
+    @gen.coroutine
+    def update(self, x, who=None):
+        now = time()
+        old_next = self.next
+        self.next = max(now, self.next) + self.interval
+        if now < old_next:
+            yield gen.sleep(old_next - now)
+        yield self.emit(x)
+
+
+class buffer(Stream):
+    def __init__(self, n, child, loop=None):
+        self.queue = Queue(maxsize=n)
+
+        Stream.__init__(self, child, loop=loop)
+
+        self.loop.add_callback(self.cb)
+
+    def update(self, x, who=None):
+        return self.queue.put(x)
+
+    @gen.coroutine
+    def cb(self):
+        while True:
+            x = yield self.queue.get()
+            yield self.emit(x)
+
+
+class zip(Stream):
+    def __init__(self, *children, **kwargs):
+        self.maxsize = kwargs.pop('maxsize', 10)
+        self.buffers = [deque() for _ in children]
+        self.condition = Condition()
+        Stream.__init__(self, children=children)
+
+    def update(self, x, who=None):
+        L = self.buffers[self.children.index(who)]
+        L.append(x)
+        if len(L) == 1 and all(self.buffers):
+            tup = tuple(buf.popleft() for buf in self.buffers)
+            self.condition.notify_all()
+            if tup and hasattr(tup[0], '__stream_merge__'):
+                tup = tup[0].__stream_merge__(*tup[1:])
+            return self.emit(tup)
+        elif len(L) > self.maxsize:
+            return self.condition.wait()
+
+
+class combine_latest(Stream):
```

Comment:
Need to make certain that we don't mismatch documents here. This may need to hold onto documents (eg stops and events) which need to be emitted together or replicated.